framework/src/meshgenerators/Poly2TriMeshGenerator.C

```diff
@@ -0,0 +1,378 @@
+
+    auto _subdomain_ids = MooseMeshUtils::getSubdomainIDs(*mesh, _subdomain_names);
+
+    // Check that the requested subdomains exist in the mesh
+    std::set<SubdomainID> subdomains;
+    mesh->subdomain_ids(subdomains);
+
+    for (auto i : index_range(_subdomain_ids))
+    {
+      if (_subdomain_ids[i] == Moose::INVALID_BLOCK_ID || !subdomains.count(_subdomain_ids[i]))
+        paramError(
+            "input_subdomain_names", _subdomain_names[i], " was not found in the boundary mesh");
+
+      bdy_ids.insert(_subdomain_ids[i]);
+    }
+  }
+
+  if (!bdy_ids.empty())
+    poly2tri.set_outer_boundary_ids(bdy_ids);
+
+  poly2tri.set_interpolate_boundary_points(_add_nodes_per_boundary_segment);
+  poly2tri.set_refine_boundary_allowed(_refine_bdy);
+  poly2tri.set_verify_hole_boundaries(_verify_holes);
+
+  poly2tri.desired_area() = _desired_area;
+  poly2tri.minimum_angle() = 0; // Not yet supported
+  poly2tri.smooth_after_generating() = _smooth_tri;
+
+  if (_desired_area_func != "")
+  {
+    // poly2tri will clone this so it's fine going out of scope
+    ParsedFunction<Real> area_func{_desired_area_func};
+    poly2tri.set_desired_area_function(&area_func);
+  }
+
+  std::vector<TriangulatorInterface::MeshedHole> meshed_holes;
+  std::vector<TriangulatorInterface::Hole *> triangulator_hole_ptrs;
+
+  // Make sure pointers here aren't invalidated by a resize
+  meshed_holes.reserve(_hole_ptrs.size());
+  for (auto hole_i : index_range(_hole_ptrs))
+  {
+    meshed_holes.emplace_back(*_hole_ptrs[hole_i]->get());
+    if (hole_i < _refine_holes.size())
+      meshed_holes.back().set_refine_boundary_allowed(_refine_holes[hole_i]);
+
+    triangulator_hole_ptrs.push_back(&meshed_holes.back());
+  }
+
+  if (!triangulator_hole_ptrs.empty())
+    poly2tri.attach_hole_list(&triangulator_hole_ptrs);
+
+  poly2tri.triangulate();
+
+  if (_smooth_tri || _output_subdomain_id)
+    for (auto elem : mesh->element_ptr_range())
+    {
+      mooseAssert(elem->type() == TRI3, "Unexpected non-Tri3 found in triangulation");
+
+      elem->subdomain_id() = _output_subdomain_id;
+
+      // I do not trust Laplacian mesh smoothing not to invert
+      // elements near reentrant corners.  Eventually we'll add better
+      // smoothing options, but even those might have failure cases.
+      // Better to always do extra tests here than to ever let users
+      // try to run on a degenerate mesh.
+      if (_smooth_tri)
+      {
+        auto cross_prod = (elem->point(1) - elem->point(0)).cross(elem->point(2) - elem->point(0));
+
+        if (cross_prod(2) <= 0)
+          mooseError("Inverted element found in triangulation.\n"
+                     "Laplacian smoothing can create these at reentrant corners; disable it?");
+      }
+    }
+
+  // Assign new subdomain name, if provided
+  if (isParamValid("output_subdomain_name"))
+    mesh->subdomain_name(_output_subdomain_id) = getParam<SubdomainName>("output_subdomain_name");
```

Comment:
yeah, I think this is redundant. I would rather reduce capability and have someone use a rename generator if desired

Reply:
That wouldn't work in general.  This generator takes arbitrary input meshes.  Any "default" name (or especially id) we might try to use here would then make that name unusable by the inputs, because by the time you got to a rename generator afterward you'd have both subdomains already mixed together.

I guess you could have users look out for those cases, rename the input name to something non-conflicting preemptively, then rename it back (as well as doing the rename for the subdomain added here) after ... but isn't this much more user friendly here?

Reply:
Yeah... I suppose this is okay. I think this applies to my conversation about `SubdomainName` as well, hmm?

Reply:
Hmmm... actually, I think in both cases as long as users can pick their own names, it wouldn't hurt to cut down the parameter list and make them get autogenerated (or pre-existing if the name's already there) ids.  If they really need name "Fourteen" to correspond to id 14 then ... well, it looks like MOOSE just doesn't want to support that in general, does it?  RenameBlockGenerator has deprecated new_block_id+new_block_name in favor of plain new_block?  If that's the case then I'm fine not supporting it either.

Reply:
Man, this is more of a pain than I had expected.  Whether or not MOOSE wants to care about specific ids, exodiff sure cares.  On the bright side, I caught a related bug in the new feature while investigating changing it.

Reply:
What's the verdict on this?

Reply:
Summarizing yesterday's Slack conversation for the benefit of others:

The verdict on "using a rename generator to get rid of code here" was "no", for both boundary and subdomain renaming.  We can't do that with another MeshGenerator as an input, because it can't be done until after triangulation is done so the newly triangulated elements get renamed/renumbered.  We can't do it with another MeshGenerator as an output, because it can't be done after hole stitching is done since that's already too late to avoid any conflicts.  We *might* be able to create a local MeshGenerator, not registered with MooseApp or anything .. but the setup for that is complicated enough to not save us code in the end, and we're not sure if that's a safe way to operate anyway, so I'm leaving the renumbering manual.

And regarding the other half of the discussion here:

The verdict on "cut down the parameter list" was "yes", and that's now done.  Though ... we might need to revisit that later, if we might run into conflicts from manually distinct names being assigned to the same non-distinct id number (because those were in separate input meshes, and the mapping is mesh-local not global).  I think we already have the same problem in StitchedMeshGenerator, too, and I'm not even sure what the right resolution to it would be.  If a user loads one Exodus input where subdomain id 1 was the "foo" subdomain, and another where it was the "bar" subdomain, and tries to stitch them both into the same output ... at the moment I think id 1 would just get renamed to "bar", but what would we want to happen instead?  Renumber a subdomain automatically without warning, I guess?  That would be anathema in libMesh, but the modern MOOSE idiom seems to be "only subdomain names are meaningful, and ids are just a way to temporarily cache a name", so it seems like the right solution to put in at the MOOSE level.

Reply:
I'm content with the above.